pkg/logutils/log.go

```diff
@@ -0,0 +1,190 @@
+)
+
+// FileOutputOption supports the configuration for log file
+type FileOutputOption struct {
+	Filename   string
+	MaxSize    int
+	MaxAge     int
+	MaxBackups int
+}
+
+// LogMode is a type help you choose the log output mode, which supports "stderr","stdout","file".
+type LogMode uint32
+
+const (
+	_             LogMode = 1 << (4 - 1 - iota)
+	OUTPUT_STDOUT         // 0100
+	OUTPUT_STDERR         // 0010
+	OUTPUT_FILE           // 0001
+)
+
+// LogMode is a type help you choose the log level.
+type LogLevel string
+
+var (
+	LogDebug LogLevel = "debug"
+	LogInfo  LogLevel = "info"
+	LogWarn  LogLevel = "warn"
+	LogError LogLevel = "error"
+	LogPanic LogLevel = "panic"
+	LogFatal LogLevel = "fatal"
+)
+
+// NewLoggerWithOption provides the ability to custom log with options.
+// You can choose 'output format', 'output mode' and decide to use 'time prefix', 'function caller suffix'
+// 'log level prefix' or not.
+// If you choose 'file output mode', you can use 'stdout and file' or 'stderr and file' together.
+func NewLoggerWithOption(format LogFormat, outputMode LogMode, fileOutputOption *FileOutputOption,
+	addTimePrefix, addLogLevelPrefix, addFuncCallerSuffix bool, logLevel LogLevel) (*zap.Logger, error) {
+
+	var fileLoggerConf lumberjack.Logger
+	if fileOutputOption != nil {
+		fileLoggerConf.Filename = fileOutputOption.Filename
+		fileLoggerConf.MaxSize = fileOutputOption.MaxSize
+		fileLoggerConf.MaxAge = fileOutputOption.MaxAge
+		fileLoggerConf.MaxBackups = fileOutputOption.MaxBackups
+	}
+
+	var ws zapcore.WriteSyncer
+	switch outputMode {
+	case OUTPUT_FILE:
+		ws = zapcore.AddSync(&fileLoggerConf)
+	case OUTPUT_STDOUT:
+		ws = zapcore.AddSync(os.Stdout)
+	case OUTPUT_STDERR:
+		ws = zapcore.AddSync(os.Stderr)
+	case OUTPUT_STDOUT | OUTPUT_FILE:
+		ws = zapcore.NewMultiWriteSyncer(zapcore.AddSync(os.Stdout), zapcore.AddSync(&fileLoggerConf))
+	case OUTPUT_STDERR | OUTPUT_FILE:
+		ws = zapcore.NewMultiWriteSyncer(zapcore.AddSync(os.Stderr), zapcore.AddSync(&fileLoggerConf))
+	case OUTPUT_STDOUT | OUTPUT_STDERR:
+		return nil, fmt.Errorf("log output mode can't set to stdout with stderr together.")
+	}
+
+	// set zap encoder configuration
+	encoderConfig := zap.NewProductionEncoderConfig()
+	if addTimePrefix {
+		encoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+	} else {
+		encoderConfig.EncodeTime = nil
+	}
+
+	if addFuncCallerSuffix {
+		encoderConfig.EncodeCaller = zapcore.ShortCallerEncoder
+	} else {
+		encoderConfig.EncodeCaller = nil
+	}
+
+	if addLogLevelPrefix {
+		encoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
+	} else {
+		encoderConfig.EncodeLevel = nil
+	}
+
+	logger := zap.New(zapcore.NewCore(
+		convertToZapFormatEncoder(format, encoderConfig),
+		ws,
+		zap.NewAtomicLevelAt(ConvertToZapLevel(logLevel)),
+	), zap.AddCaller())
+	return logger, nil
+}
+
+// NewLogger create  Logger instance with default configuration for 'stdout' usage, it's JsonLogFormat.
+func NewLogger() (err error) {
```

Comment:
it initial Logger in this package, so the function name is not proper ,  it does not return new one ,
InitStdLogger()  make much sense